In [1]:
%load_ext sql

# put a folder and DB credential files at HOME directory

import os
homedir = os.getcwd()
cred_path = os.path.join(homedir, 'db_cred')



# add a 'cred_path' for interpreter to search
import sys
sys.path.append(cred_path)



# import DB credentials from 'gpdb_credentials.py' dictionary file.
from gpdb_credentials import dvdrental_db



# parsing DB credentials and connect to Greenplum using %sql $connection_string

username = dvdrental_db['Username']
password = dvdrental_db['Password']
host = dvdrental_db['Host']
port = dvdrental_db['Port']
database = dvdrental_db['Database']

connection_string = 'postgresql://{user}:{password}@{host}:{port}/{db}'.format(
user=username,
password=password,
host=host,
port=port,
db=database)

%sql $connection_string

'Connected: myuser@dvdrental'

# 문제 1

> PAYMENT 테이블에서 `단일거래의 AMOUNT 액수`가 가장 많은 고객들의 CUSTOMER_ID를 추출하시오. 단, `CUSTOMER_ID의 값은 '유일'` 해야 한다.

- 키워드는 '단일거래 amount 액수' 와 'customer_id 값이 유일해야 한다' 는 점이다.
- customer_id는 고객 1명마다 주어지는 고유 id 이다. 
- 동일 고객이 여러건을 결제 할 수 는 있다. 
- 하지만, 그중 가장 높은 금액의 결제를 한 1건의 지불내역이 있을 것이다.
- 물론 동일 고객이 결제한 최고 금액이 중복될 수도 있다.
    - 그래서 distinct를 사용해서 중복제거를 해준다.


- 그러므로, customer_id 중에 단일거래 amount 가 가장 큰값 을 찾는다. = sub-query를 사용하면 된다.
- `단일 거래 amount 액수` = 1건 거래시 가장 많은 액수를 결제한 amount 
    - order by 와 limit을 통해 해결 가능 하다.

1) 가장 높은 수치를 가진 amount를 확인 -> suq query에 들어갈 내용

In [2]:
%%sql

select amount
from payment
order by amount desc
limit 1;

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
1 rows affected.


[(Decimal('11.99'),)]

2) PAYMENT 테이블에서 가장 큰 AMOUNT를 가진 CUSTOMER_ID를 구한뒤 중복제거를 한다.

Tip)

서브쿼리가 들어가면서 부터는 alias를 적극 활용해주는 것이 좋다.

동일한 테이블의 동일한 컬럼에서 조건을 부여하기 때문에 작성자 외의 제3자가 보았을때도 한눈에 이해가 쉽게 하기 위함

In [3]:
%%sql

SELECT
	DISTINCT A.CUSTOMER_ID -- DISTINCT로 중복 제거
FROM
	PAYMENT A -- 메인 쿼리는 A라 지칭
WHERE
	A.AMOUNT = -- A 테이블(메인쿼리) 의 AMOUNT
    (
		SELECT
			K.AMOUNT -- K 테이블 (서브쿼리)의 AMOUNT
		FROM
			PAYMENT K -- 서브 쿼리는 K라 지칭
		ORDER BY
			K.AMOUNT DESC
		LIMIT 1 -- LIMIT절로 1개의 값만 추출
    );

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
8 rows affected.


[(116,), (204,), (13,), (195,), (591,), (237,), (592,), (362,)]

subquery에서 추출되는 집합의 값이 limit절에 의해 딱 1개 이기 때문에 '=' 연산자를 사용할 수 있다.

# 문제 2

> 고객들에게 단체 이메일을 전송하고자 한다. CUSTOMER 테이블에서 고객의 EMAIL주소를 추출하고, 이메일 형식에 맞지 않는 이메일 주소는 제외시켜라. 
단, 이메일 형식은 '@'가 존재해야 하지만, '@'로 시작해서는 안되고, '@'로 끝나서도 안된다.

- LIKE절을 활용해서 3가지 조건을 동시에 만족하는 집합을 추출하면 된다.
    - 이메일에 '@'반드시 들어간다 = WHERE EMAIL LIKE '%@%'= 시작과 끝은 어떤 문자가 와도 되지만 중간에 @가 꼭들어가야 한다.
    - 이메일이 '@'부터 시작하면 안된다 = WHERE EMAIL NOT LIKE '@%'
    - 이메일이 '@'로 끝나면 안된다 = = WHERE EMAIL NOT LIKE '%@'

In [4]:
%%sql

SELECT
	EMAIL
FROM
	CUSTOMER
WHERE
	EMAIL NOT LIKE '@%' -- '@'로 시작하면 안된다.
	AND EMAIL NOT LIKE '%@' -- '@'로 끝나면 안된다.
	AND EMAIL LIKE '%@%'; -- '@'가 반드시 들어가야 한다.

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
599 rows affected.


[('jared.ely@sakilacustomer.org',),
 ('mary.smith@sakilacustomer.org',),
 ('patricia.johnson@sakilacustomer.org',),
 ('linda.williams@sakilacustomer.org',),
 ('barbara.jones@sakilacustomer.org',),
 ('elizabeth.brown@sakilacustomer.org',),
 ('jennifer.davis@sakilacustomer.org',),
 ('maria.miller@sakilacustomer.org',),
 ('susan.wilson@sakilacustomer.org',),
 ('margaret.moore@sakilacustomer.org',),
 ('dorothy.taylor@sakilacustomer.org',),
 ('lisa.anderson@sakilacustomer.org',),
 ('nancy.thomas@sakilacustomer.org',),
 ('karen.jackson@sakilacustomer.org',),
 ('betty.white@sakilacustomer.org',),
 ('helen.harris@sakilacustomer.org',),
 ('sandra.martin@sakilacustomer.org',),
 ('donna.thompson@sakilacustomer.org',),
 ('carol.garcia@sakilacustomer.org',),
 ('ruth.martinez@sakilacustomer.org',),
 ('sharon.robinson@sakilacustomer.org',),
 ('michelle.clark@sakilacustomer.org',),
 ('laura.rodriguez@sakilacustomer.org',),
 ('sarah.lewis@sakilacustomer.org',),
 ('kimberly.lee@sakilacustomer.org',),
 ('deborah.walker@sakilacustomer.org',),
 ('jessica.hall@sakilacustomer.org',),
 ('shirley.allen@sakilacustomer.org',),
 ('cynthia.young@sakilacustomer.org',),
 ('angela.hernandez@sakilacustomer.org',),
 ('melissa.king@sakilacustomer.org',),
 ('brenda.wright@sakilacustomer.org',),
 ('amy.lopez@sakilacustomer.org',),
 ('anna.hill@sakilacustomer.org',),
 ('rebecca.scott@sakilacustomer.org',),
 ('virginia.green@sakilacustomer.org',),
 ('kathleen.adams@sakilacustomer.org',),
 ('pamela.baker@sakilacustomer.org',),
 ('martha.gonzalez@sakilacustomer.org',),
 ('debra.nelson@sakilacustomer.org',),
 ('amanda.carter@sakilacustomer.org',),
 ('stephanie.mitchell@sakilacustomer.org',),
 ('carolyn.perez@sakilacustomer.org',),
 ('christine.roberts@sakilacustomer.org',),
 ('marie.turner@sakilacustomer.org',),
 ('janet.phillips@sakilacustomer.org',),
 ('catherine.campbell@sakilacustomer.org',),
 ('frances.parker@sakilacustomer.org',),
 ('ann.evans@sakilacustomer.org',),
 ('joyce.edwards@sakilacustomer.org',),
 ('diane.collins@sakilacustomer.org',),
 ('alice.stewart@sakilacustomer.org',),
 ('julie.sanchez@sakilacustomer.org',),
 ('heather.morris@sakilacustomer.org',),
 ('teresa.rogers@sakilacustomer.org',),
 ('doris.reed@sakilacustomer.org',),
 ('gloria.cook@sakilacustomer.org',),
 ('evelyn.morgan@sakilacustomer.org',),
 ('jean.bell@sakilacustomer.org',),
 ('cheryl.murphy@sakilacustomer.org',),
 ('mildred.bailey@sakilacustomer.org',),
 ('katherine.rivera@sakilacustomer.org',),
 ('joan.cooper@sakilacustomer.org',),
 ('ashley.richardson@sakilacustomer.org',),
 ('judith.cox@sakilacustomer.org',),
 ('rose.howard@sakilacustomer.org',),
 ('janice.ward@sakilacustomer.org',),
 ('kelly.torres@sakilacustomer.org',),
 ('nicole.peterson@sakilacustomer.org',),
 ('judy.gray@sakilacustomer.org',),
 ('christina.ramirez@sakilacustomer.org',),
 ('kathy.james@sakilacustomer.org',),
 ('theresa.watson@sakilacustomer.org',),
 ('beverly.brooks@sakilacustomer.org',),
 ('denise.kelly@sakilacustomer.org',),
 ('tammy.sanders@sakilacustomer.org',),
 ('irene.price@sakilacustomer.org',),
 ('jane.bennett@sakilacustomer.org',),
 ('lori.wood@sakilacustomer.org',),
 ('rachel.barnes@sakilacustomer.org',),
 ('marilyn.ross@sakilacustomer.org',),
 ('andrea.henderson@sakilacustomer.org',),
 ('kathryn.coleman@sakilacustomer.org',),
 ('louise.jenkins@sakilacustomer.org',),
 ('sara.perry@sakilacustomer.org',),
 ('anne.powell@sakilacustomer.org',),
 ('jacqueline.long@sakilacustomer.org',),
 ('wanda.patterson@sakilacustomer.org',),
 ('bonnie.hughes@sakilacustomer.org',),
 ('julia.flores@sakilacustomer.org',),
 ('ruby.washington@sakilacustomer.org',),
 ('lois.butler@sakilacustomer.org',),
 ('tina.simmons@sakilacustomer.org',),
 ('phyllis.foster@sakilacustomer.org',),
 ('norma.gonzales@sakilacustomer.org',),
 ('paula.bryant@sakilacustomer.org',),
 ('diana.alexander@sakilacustomer.org',),
 ('annie.russell@sakilacustomer.org',),
 ('lillian.griffin@sakilacustomer.org',),
 ('emily.diaz@sakilacustomer.org',),
 ('robi

- 그냥 EMAIL 전체 데이터를 출력해도 599 rows 인걸로 보아 기본적인 데이터셋 자체가 잘못된 형식으로 입력된 데이터는 없는 것 같다.
- 실제 이메일을 보내보기 전까지는 입력된 메일 주소가 실제로 있는 것인지 까지는 SQL로 파악이 어렵지만, 적어도 올바른 형식을 가진 이메일 들만 추출해서 추려낼 수는 있다.

# 문제 3

>ACTOR 테이블의 성(LAST_NAME)이 Jo 로 시작하는 사람의 ID값이 가장 낮은 '한사람' 의 ID값, 이름(FIRST_NAME), 성(LAST_NAME)을 알려주세요

- actor 테이블의 id = actor_id 를 의미
- 조건 1 : last_name이 'Jo'로 시작하는 데이터
- 조건 2 : 조건 1에 해당 하는 사람중 actor_id 값이 가장 낮은 데이터 = order by actor_id asc를 했을때 맨 위의 값
- 조건 3 : 조건 1과 2를 모두 만족하는 데이터의 actor_id, first_name, last_name을 출력

조건 1을 만족하는 데이터 찾기

In [5]:
%%sql

SELECT
	ACTOR_ID,
	FIRST_NAME,
	LAST_NAME
FROM
	ACTOR
WHERE
	LAST_NAME LIKE 'Jo%';

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
5 rows affected.


[(8, 'Matthew', 'Johansson'),
 (43, 'Kirk', 'Jovovich'),
 (64, 'Ray', 'Johansson'),
 (82, 'Woody', 'Jolie'),
 (146, 'Albert', 'Johansson')]

조건 2와 3을 만족하는 데이터 찾기

In [6]:
%%sql

SELECT
	ACTOR_ID,
	FIRST_NAME,
	LAST_NAME
FROM
	ACTOR
WHERE
	LAST_NAME LIKE 'Jo%' -- Jo로 시작
ORDER BY
	ACTOR_ID -- asc 정렬
LIMIT 1; -- limit으로 상위 1개의 값만 출력

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
1 rows affected.


[(8, 'Matthew', 'Johansson')]

# 문제 4

>FILM 테이블을 이용해서, FILM 테이블의 아이디 값이 1 ~ 10 사이에 있는 모든 컬럼을 확인해주세요.

between 을 활용

In [7]:
%%sql

SELECT *
FROM
	FILM
WHERE
	FILM_ID BETWEEN 1 AND 10;

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
10 rows affected.


[(1, 'Academy Dinosaur', 'A Epic Drama of a Feminist And a Mad Scientist who must Battle a Teacher in The Canadian Rockies', 2006, 1, 6, Decimal('0.99'), 86, Decimal('20.99'), 'PG', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes', 'Behind the Scenes'], "'academi':1 'battl':15 'canadian':20 'dinosaur':2 'drama':5 'epic':4 'feminist':8 'mad':11 'must':14 'rocki':21 'scientist':12 'teacher':17"),
 (2, 'Ace Goldfinger', 'A Astounding Epistle of a Database Administrator And a Explorer who must Find a Car in Ancient China', 2006, 1, 3, Decimal('4.99'), 48, Decimal('12.99'), 'G', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'ace':1 'administr':9 'ancient':19 'astound':4 'car':17 'china':20 'databas':8 'epistl':5 'explor':12 'find':15 'goldfing':2 'must':14"),
 (3, 'Adaptation Holes', 'A Astounding Reflection of a Lumberjack And a Car who must Sink a Lumberjack in A Baloon Factory', 2006, 1, 7, Decimal('2.99'), 50, Decimal('18.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'adapt':1 'astound':4 'baloon':19 'car':11 'factori':20 'hole':2 'lumberjack':8,16 'must':13 'reflect':5 'sink':14"),
 (4, 'Affair Prejudice', 'A Fanciful Documentary of a Frisbee And a Lumberjack who must Chase a Monkey in A Shark Tank', 2006, 1, 5, Decimal('2.99'), 117, Decimal('26.99'), 'G', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Commentaries', 'Behind the Scenes'], "'affair':1 'chase':14 'documentari':5 'fanci':4 'frisbe':8 'lumberjack':11 'monkey':16 'must':13 'prejudic':2 'shark':19 'tank':20"),
 (5, 'African Egg', 'A Fast-Paced Documentary of a Pastry Chef And a Dentist who must Pursue a Forensic Psychologist in The Gulf of Mexico', 2006, 1, 6, Decimal('2.99'), 130, Decimal('22.99'), 'G', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes'], "'african':1 'chef':11 'dentist':14 'documentari':7 'egg':2 'fast':5 'fast-pac':4 'forens':19 'gulf':23 'mexico':25 'must':16 'pace':6 'pastri':10 'psychologist':20 'pursu':17"),
 (6, 'Agent Truman', 'A Intrepid Panorama of a Robot And a Boy who must Escape a Sumo Wrestler in Ancient China', 2006, 1, 3, Decimal('2.99'), 169, Decimal('17.99'), 'PG', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes'], "'agent':1 'ancient':19 'boy':11 'china':20 'escap':14 'intrepid':4 'must':13 'panorama':5 'robot':8 'sumo':16 'truman':2 'wrestler':17"),
 (7, 'Airplane Sierra', 'A Touching Saga of a Hunter And a Butler who must Discover a Butler in A Jet Boat', 2006, 1, 6, Decimal('4.99'), 62, Decimal('28.99'), 'PG-13', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'airplan':1 'boat':20 'butler':11,16 'discov':14 'hunter':8 'jet':19 'must':13 'saga':5 'sierra':2 'touch':4"),
 (8, 'Airport Pollock', 'A Epic Tale of a Moose And a Girl who must Confront a Monkey in Ancient India', 2006, 1, 6, Decimal('4.99'), 54, Decimal('15.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'airport':1 'ancient':18 'confront':14 'epic':4 'girl':11 'india':19 'monkey':16 'moos':8 'must':13 'pollock':2 'tale':5"),
 (9, 'Alabama Devil', 'A Thoughtful Panorama of a Database Administrator And a Mad Scientist who must Outgun a Mad Scientist in A Jet Boat', 2006, 1, 3, Decimal('2.99'), 114, Decimal('21.99'), 'PG-13', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'administr':9 'alabama':1 'boat':23 'databas':8 'devil':2 'jet':22 'mad':12,18 'must':15 'outgun':16 'panorama':5 'scientist':13,19 'thought':4"),
 (10, 'Aladdin Calendar', 'A Action-Packed Tale of a Man And a Lumberjack who must Reach a Feminist in Ancient China', 2006, 1, 6, Decimal('4.99'), 63, Decimal('24.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'action':5 'action-pack':4 'aladdin':1 'ancient':20 'calendar':2 'china':21 'feminist':18 'lumberjack':13 'man':10 'must':15 'pack':6 '

order by와 limit을 활용

In [8]:
%%sql
SELECT *
FROM
	FILM
ORDER BY
	FILM_ID
LIMIT 10;

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
10 rows affected.


[(1, 'Academy Dinosaur', 'A Epic Drama of a Feminist And a Mad Scientist who must Battle a Teacher in The Canadian Rockies', 2006, 1, 6, Decimal('0.99'), 86, Decimal('20.99'), 'PG', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes', 'Behind the Scenes'], "'academi':1 'battl':15 'canadian':20 'dinosaur':2 'drama':5 'epic':4 'feminist':8 'mad':11 'must':14 'rocki':21 'scientist':12 'teacher':17"),
 (2, 'Ace Goldfinger', 'A Astounding Epistle of a Database Administrator And a Explorer who must Find a Car in Ancient China', 2006, 1, 3, Decimal('4.99'), 48, Decimal('12.99'), 'G', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'ace':1 'administr':9 'ancient':19 'astound':4 'car':17 'china':20 'databas':8 'epistl':5 'explor':12 'find':15 'goldfing':2 'must':14"),
 (3, 'Adaptation Holes', 'A Astounding Reflection of a Lumberjack And a Car who must Sink a Lumberjack in A Baloon Factory', 2006, 1, 7, Decimal('2.99'), 50, Decimal('18.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'adapt':1 'astound':4 'baloon':19 'car':11 'factori':20 'hole':2 'lumberjack':8,16 'must':13 'reflect':5 'sink':14"),
 (4, 'Affair Prejudice', 'A Fanciful Documentary of a Frisbee And a Lumberjack who must Chase a Monkey in A Shark Tank', 2006, 1, 5, Decimal('2.99'), 117, Decimal('26.99'), 'G', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Commentaries', 'Behind the Scenes'], "'affair':1 'chase':14 'documentari':5 'fanci':4 'frisbe':8 'lumberjack':11 'monkey':16 'must':13 'prejudic':2 'shark':19 'tank':20"),
 (5, 'African Egg', 'A Fast-Paced Documentary of a Pastry Chef And a Dentist who must Pursue a Forensic Psychologist in The Gulf of Mexico', 2006, 1, 6, Decimal('2.99'), 130, Decimal('22.99'), 'G', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes'], "'african':1 'chef':11 'dentist':14 'documentari':7 'egg':2 'fast':5 'fast-pac':4 'forens':19 'gulf':23 'mexico':25 'must':16 'pace':6 'pastri':10 'psychologist':20 'pursu':17"),
 (6, 'Agent Truman', 'A Intrepid Panorama of a Robot And a Boy who must Escape a Sumo Wrestler in Ancient China', 2006, 1, 3, Decimal('2.99'), 169, Decimal('17.99'), 'PG', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes'], "'agent':1 'ancient':19 'boy':11 'china':20 'escap':14 'intrepid':4 'must':13 'panorama':5 'robot':8 'sumo':16 'truman':2 'wrestler':17"),
 (7, 'Airplane Sierra', 'A Touching Saga of a Hunter And a Butler who must Discover a Butler in A Jet Boat', 2006, 1, 6, Decimal('4.99'), 62, Decimal('28.99'), 'PG-13', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'airplan':1 'boat':20 'butler':11,16 'discov':14 'hunter':8 'jet':19 'must':13 'saga':5 'sierra':2 'touch':4"),
 (8, 'Airport Pollock', 'A Epic Tale of a Moose And a Girl who must Confront a Monkey in Ancient India', 2006, 1, 6, Decimal('4.99'), 54, Decimal('15.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'airport':1 'ancient':18 'confront':14 'epic':4 'girl':11 'india':19 'monkey':16 'moos':8 'must':13 'pollock':2 'tale':5"),
 (9, 'Alabama Devil', 'A Thoughtful Panorama of a Database Administrator And a Mad Scientist who must Outgun a Mad Scientist in A Jet Boat', 2006, 1, 3, Decimal('2.99'), 114, Decimal('21.99'), 'PG-13', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'administr':9 'alabama':1 'boat':23 'databas':8 'devil':2 'jet':22 'mad':12,18 'must':15 'outgun':16 'panorama':5 'scientist':13,19 'thought':4"),
 (10, 'Aladdin Calendar', 'A Action-Packed Tale of a Man And a Lumberjack who must Reach a Feminist in Ancient China', 2006, 1, 6, Decimal('4.99'), 63, Decimal('24.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'action':5 'action-pack':4 'aladdin':1 'ancient':20 'calendar':2 'china':21 'feminist':18 'lumberjack':13 'man':10 'must':15 'pack':6 '

어느 방식을 사용해도 결과는 같다.

# 문제 5

>ADDRESS 테이블을 이용해서, 우편번호(POSTAL_CODE) 값의 두번째 글자가 1인 우편번호의 ADDRESS_ID, ADDRESS, DISTRICT, POSTAL_CODE 컬럼 값을 확인해주세요.

- LIKE절과 와일드카드(%) 활용해 쉽게 해결 가능하다.
- 또한 전체 문자열에서 특정 위치의 값을 추출하는 substring과 비교연산자('=')을 활용해서 풀수도 있다.

- LIKE절과 와일드카드(%) 활용

In [9]:
%%sql
SELECT
	ADDRESS_ID,
	ADDRESS,
	DISTRICT,
	POSTAL_CODE,
	SUBSTRING(POSTAL_CODE,2,1) TEST1
FROM
	ADDRESS
WHERE
	POSTAL_CODE LIKE '_1%';

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
54 rows affected.


[(21, '270 Toulon Boulevard', 'Kalmykia', '81766', '1'),
 (24, '1688 Okara Way', 'Nothwest Border Prov', '21954', '1'),
 (46, '1632 Bislig Avenue', 'Nonthaburi', '61117', '1'),
 (56, '939 Probolinggo Loop', 'Galicia', '4166', '1'),
 (69, '698 Otsu Street', 'Cayenne', '71110', '1'),
 (71, '1586 Guaruj Place', 'Hunan', '5135', '1'),
 (101, '1599 Plock Drive', 'Tete', '71986', '1'),
 (103, '588 Vila Velha Manor', 'Kyongsangbuk', '51540', '1'),
 (114, '804 Elista Drive', 'Hubei', '61069', '1'),
 (140, '788 Atinsk Street', 'Karnataka', '81691', '1'),
 (141, '1749 Daxian Place', 'Gelderland', '11044', '1'),
 (159, '185 Novi Sad Place', 'Bern', '41778', '1'),
 (160, '1367 Yantai Manor', 'Ondo & Ekiti', '21294', '1'),
 (163, '1440 Compton Place', 'North Austria', '81037', '1'),
 (164, '1623 Baha Blanca Manor', 'Moskova', '81511', '1'),
 (207, '1688 Nador Lane', 'Sulawesi Utara', '61613', '1'),
 (220, '1201 Qomsheh Manor', 'Gois', '21464', '1'),
 (236, '885 Yingkou Manor', 'Kaduna', '31390', '1'),
 (249, '300 Junan Street', 'Kyonggi', '81314', '1'),
 (268, '1831 Nam Dinh Loop', 'National Capital Reg', '51990', '1'),
 (276, '1675 Xiangfan Manor', 'Tamil Nadu', '11763', '1'),
 (287, '1405 Chisinau Place', 'Ponce', '8160', '1'),
 (296, '1936 Cuman Avenue', 'Virginia', '61195', '1'),
 (298, '44 Najafabad Way', 'Baskimaa', '61391', '1'),
 (318, '1774 Yaound Place', 'Hubei', '91400', '1'),
 (335, '587 Benguela Manor', 'Illinois', '91590', '1'),
 (337, '1838 Tabriz Lane', 'Dhaka', '1195', '1'),
 (343, '1443 Mardan Street', 'Western Cape', '31483', '1'),
 (347, '1293 Nam Dinh Way', 'Roraima', '71583', '1'),
 (360, '619 Hunuco Avenue', 'Shimane', '81508', '1'),
 (364, '1623 Kingstown Drive', 'Buenos Aires', '91299', '1'),
 (374, '433 Florencia Street', 'Chihuahua', '91330', '1'),
 (393, '717 Changzhou Lane', 'Southern Tagalog', '21615', '1'),
 (405, '530 Lausanne Lane', 'Texas', '11067', '1'),
 (407, '1346 Mysore Drive', 'Bretagne', '61507', '1'),
 (414, '1936 Lapu-Lapu Parkway', 'Bauchi & Gombe', '7122', '1'),
 (430, '355 Vitria de Santo Anto Way', 'Oaxaca', '81758', '1'),
 (439, '1351 Aparecida de Goinia Parkway', 'Northern Mindanao', '41775', '1'),
 (454, '956 Nam Dinh Manor', 'Kerman', '21872', '1'),
 (457, '535 Ahmadnagar Manor', 'Abu Dhabi', '41136', '1'),
 (472, '1447 Chatsworth Place', 'Chihuahua', '41545', '1'),
 (480, '421 Yaound Street', 'Sumy', '11363', '1'),
 (482, '808 Naala-Porto Parkway', 'England', '41060', '1'),
 (487, '1785 So Bernardo do Campo Street', 'Veracruz', '71182', '1'),
 (489, '1297 Alvorada Parkway', 'Ningxia', '11839', '1'),
 (514, '1747 Rustenburg Place', 'Bihar', '51369', '1'),
 (525, '1949 Sanya Street', 'Gumma', '61244', '1'),
 (527, '1993 0 Loop', 'Liaoning', '41214', '1'),
 (528, '1176 Southend-on-Sea Manor', 'Southern Tagalog', '81651', '1'),
 (532, '1427 Tabuk Place', 'Florida', '31342', '1'),
 (548, '1658 Cuman Loop', 'Sumatera Selatan', '51309', '1'),
 (559, '1917 Kumbakonam Parkway', 'Vojvodina', '11892', '1'),
 (593, '1402 Zanzibar Boulevard', 'Guanajuato', '71102', '1'),
 (604, '1331 Usak Boulevard', 'Vaud', '61960', '1')]

- substring과 비교연산자('=') 을 활용
    - substring(column_name,추출을 시작할 문자열의 위치번호,추출할 문자열의 개수)
    - SUBSTRING(POSTAL_CODE,2,1) 은 `POSTAL_CODE컬럼의 값에서 2번째 위치부터 1개의 문자열을 추출하겠다.` 는 의미

In [10]:
%%sql
SELECT
	ADDRESS_ID,
	ADDRESS,
	DISTRICT,
	POSTAL_CODE,
	SUBSTRING(POSTAL_CODE,2,1) TEST1
FROM
	ADDRESS
where SUBSTRING(POSTAL_CODE,2,1) = '1'

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
54 rows affected.


[(21, '270 Toulon Boulevard', 'Kalmykia', '81766', '1'),
 (24, '1688 Okara Way', 'Nothwest Border Prov', '21954', '1'),
 (46, '1632 Bislig Avenue', 'Nonthaburi', '61117', '1'),
 (56, '939 Probolinggo Loop', 'Galicia', '4166', '1'),
 (69, '698 Otsu Street', 'Cayenne', '71110', '1'),
 (71, '1586 Guaruj Place', 'Hunan', '5135', '1'),
 (101, '1599 Plock Drive', 'Tete', '71986', '1'),
 (103, '588 Vila Velha Manor', 'Kyongsangbuk', '51540', '1'),
 (114, '804 Elista Drive', 'Hubei', '61069', '1'),
 (140, '788 Atinsk Street', 'Karnataka', '81691', '1'),
 (141, '1749 Daxian Place', 'Gelderland', '11044', '1'),
 (159, '185 Novi Sad Place', 'Bern', '41778', '1'),
 (160, '1367 Yantai Manor', 'Ondo & Ekiti', '21294', '1'),
 (163, '1440 Compton Place', 'North Austria', '81037', '1'),
 (164, '1623 Baha Blanca Manor', 'Moskova', '81511', '1'),
 (207, '1688 Nador Lane', 'Sulawesi Utara', '61613', '1'),
 (220, '1201 Qomsheh Manor', 'Gois', '21464', '1'),
 (236, '885 Yingkou Manor', 'Kaduna', '31390', '1'),
 (249, '300 Junan Street', 'Kyonggi', '81314', '1'),
 (268, '1831 Nam Dinh Loop', 'National Capital Reg', '51990', '1'),
 (276, '1675 Xiangfan Manor', 'Tamil Nadu', '11763', '1'),
 (287, '1405 Chisinau Place', 'Ponce', '8160', '1'),
 (296, '1936 Cuman Avenue', 'Virginia', '61195', '1'),
 (298, '44 Najafabad Way', 'Baskimaa', '61391', '1'),
 (318, '1774 Yaound Place', 'Hubei', '91400', '1'),
 (335, '587 Benguela Manor', 'Illinois', '91590', '1'),
 (337, '1838 Tabriz Lane', 'Dhaka', '1195', '1'),
 (343, '1443 Mardan Street', 'Western Cape', '31483', '1'),
 (347, '1293 Nam Dinh Way', 'Roraima', '71583', '1'),
 (360, '619 Hunuco Avenue', 'Shimane', '81508', '1'),
 (364, '1623 Kingstown Drive', 'Buenos Aires', '91299', '1'),
 (374, '433 Florencia Street', 'Chihuahua', '91330', '1'),
 (393, '717 Changzhou Lane', 'Southern Tagalog', '21615', '1'),
 (405, '530 Lausanne Lane', 'Texas', '11067', '1'),
 (407, '1346 Mysore Drive', 'Bretagne', '61507', '1'),
 (414, '1936 Lapu-Lapu Parkway', 'Bauchi & Gombe', '7122', '1'),
 (430, '355 Vitria de Santo Anto Way', 'Oaxaca', '81758', '1'),
 (439, '1351 Aparecida de Goinia Parkway', 'Northern Mindanao', '41775', '1'),
 (454, '956 Nam Dinh Manor', 'Kerman', '21872', '1'),
 (457, '535 Ahmadnagar Manor', 'Abu Dhabi', '41136', '1'),
 (472, '1447 Chatsworth Place', 'Chihuahua', '41545', '1'),
 (480, '421 Yaound Street', 'Sumy', '11363', '1'),
 (482, '808 Naala-Porto Parkway', 'England', '41060', '1'),
 (487, '1785 So Bernardo do Campo Street', 'Veracruz', '71182', '1'),
 (489, '1297 Alvorada Parkway', 'Ningxia', '11839', '1'),
 (514, '1747 Rustenburg Place', 'Bihar', '51369', '1'),
 (525, '1949 Sanya Street', 'Gumma', '61244', '1'),
 (527, '1993 0 Loop', 'Liaoning', '41214', '1'),
 (528, '1176 Southend-on-Sea Manor', 'Southern Tagalog', '81651', '1'),
 (532, '1427 Tabuk Place', 'Florida', '31342', '1'),
 (548, '1658 Cuman Loop', 'Sumatera Selatan', '51309', '1'),
 (559, '1917 Kumbakonam Parkway', 'Vojvodina', '11892', '1'),
 (593, '1402 Zanzibar Boulevard', 'Guanajuato', '71102', '1'),
 (604, '1331 Usak Boulevard', 'Vaud', '61960', '1')]

# 문제 6

> PAYMENT 테이블을 이용해서, 고객 번호가 355에 해당하는 사람의 결제 금액이 1~3원 사이에 해당하는 모든 결제 내역을 확인해주세요.

- where절에 1개이상의 조건을 적용할때는 'and' 혹은 'or'을 통해 조건 지정을 할 수 있다.
- 이때, 서로 다른 복수 컬럼에 각각 조건을 적용할 수 도 있다.

In [11]:
%%sql

SELECT *
FROM
	PAYMENT
WHERE
	CUSTOMER_ID = '355'
	AND AMOUNT BETWEEN 1 AND 3;

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
5 rows affected.


[(17561, 355, 1, 1612, Decimal('2.99'), datetime.datetime(2007, 2, 16, 5, 20, 31, 996577)),
 (20334, 355, 2, 13821, Decimal('1.99'), datetime.datetime(2007, 3, 20, 12, 2, 13, 996577)),
 (20336, 355, 2, 15531, Decimal('2.99'), datetime.datetime(2007, 3, 23, 2, 21, 2, 996577)),
 (26172, 355, 1, 6437, Decimal('2.99'), datetime.datetime(2007, 4, 11, 22, 48, 55, 996577)),
 (26176, 355, 2, 8418, Decimal('1.99'), datetime.datetime(2007, 4, 29, 5, 22, 47, 996577))]

# 문제 7

> FILM 테이블을 이용해서, FILM의 길이가 100 ~ 120 에 해당 하거나 RENTAL 대여기간이 3 ~ 5일에 해당하는 FILM의 모든 정보를 확인해주세요.

- 조건 1) film의 길이가 100 ~ 120에 해당하는 데이터 집합 
- ~에 해당하거나 = or 조건
- 조건 2) rental 기간 (rental_duration)이 3 ~ 5일인 데이터

In [12]:
%%sql

SELECT *
FROM
	FILM
WHERE
	LENGTH BETWEEN 100 AND 120
	OR RENTAL_DURATION BETWEEN 3 AND 5 -- 662 개
LIMIT 20; -- 값이 너무 많아서 20개로 제한

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
20 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5"),
 (384, 'Grosse Wonderful', 'A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia', 2006, 1, 5, Decimal('4.99'), 49, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Behind the Scenes'], "'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2"),
 (98, 'Bright Encounters', 'A Fateful Yarn of a Lumberjack And a Feminist who must Conquer a Student in A Jet Boat', 2006, 1, 4, Decimal('4.99'), 73, Decimal('12.99'), 'PG-13', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'boat':20 'bright':1 'conquer':14 'encount':2 'fate':4 'feminist':11 'jet':19 'lumberjack':8 'must':13 'student':16 'yarn':5"),
 (2, 'Ace Goldfinger', 'A Astounding Epistle of a Database Administrator And a Explorer who must Find a Car in Ancient China', 2006, 1, 3, Decimal('4.99'), 48, Decimal('12.99'), 'G', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'ace':1 'administr':9 'ancient':19 'astound':4 'car':17 'china':20 'databas':8 'epistl':5 'explor':12 'find':15 'goldfing':2 'must':14"),
 (4, 'Affair Prejudice', 'A Fanciful Documentary of a Frisbee And a Lumberjack who must Chase a Monkey in A Shark Tank', 2006, 1, 5, Decimal('2.99'), 117, Decimal('26.99'), 'G', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Commentaries', 'Behind the Scenes'], "'affair':1 'chase':14 'documentari':5 'fanci':4 'frisbe':8 'lumberjack':11 'monkey':16 'must':13 'prejudic':2 'shark':19 'tank':20"),
 (6, 'Agent Truman', 'A Intrepid Panorama of a Robot And a Boy who must Escape a Sumo Wrestler in Ancient China', 2006, 1, 3, Decimal('2.99'), 169, Decimal('17.99'), 'PG', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes'], "'agent':1 'ancient':19 'boy':11 'china':20 'escap':14 'intrepid':4 'must':13 'panorama':5 'robot':8 'sumo':16 'truman':2 'wrestler':17"),
 (9, 'Alabama Devil', 'A Thoughtful Panorama of a Database Administrator And a Mad Scientist who must Outgun a Mad Scientist in A Jet Boat', 2006, 1, 3, Decimal('2.99'), 114, Decimal('21.99'), 'PG-13', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'administr':9 'alabama':1 'boat':23 'databas':8 'devil':2 'jet':22 'mad':12,18 'must':15 'outgun':16 'panorama':5 'scientist':13,19 'thought':4"),
 (213, 'Date Speed', 'A Touching Saga of a Composer And a Moose who must Discover a Dentist in A MySQL Convention', 2006, 1, 4, Decimal('0.99'), 104, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Commentaries'], "'compos':8 'convent':20 'date':1 'dentist':16 'discov':14 'moos':11 'must':13 'mysql':19 'saga':5 'speed':2 'touch':4"),
 (13, 'Ali Forever', 'A Action-Packed Drama of a Dentist And a Crocodile who must Battle a Feminist in The Canadian Rockies', 2006, 1, 4, Decimal('4.99'), 150, Decimal('21.99'), 'PG', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes', 'Behind the Scenes'], "'action':5 'action-pack':4 'ali':1 'battl':16 'canadian':21 'crocodil':13 'dentist':10 'drama':7 'feminist':18 'forev':2 'must':15 'pack':6 'rocki':22"),
 (15, 'Alien Center', 'A Brilliant Drama of a Cat And a Mad Scientist who must Battle a Feminist in A MySQL Convention', 2006, 1, 5, Decimal('2.99'), 46, Decimal('10.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Commentaries', 'Behind the Scenes'], "'alien':1 'battl':15 'brilliant':4 'cat':8 'center':2 'convent':21 'drama':5 'feminist':17 'mad':11 'must':14 'mysql':20 'scientist':12"),
 (17, 'Alone Trip', 'A Fast-Paced Character Study of a Composer And a Dog w

# 문제 8

> ADDRESS 테이블을 이용하여, POSTAL_CODE 값이 공백('') 이거나 35200, 17886에 해당하는 ADDRESS의 모든 정보를 확인해주세요.

- 조건 1) POSTAL_CODE 값이 공백('') or
- 조건 2) POSTAL_CODE 값이 35200 or
- 조건 3) POSTAL_CODE 값이 17886 or

In [13]:
%%sql

SELECT *
FROM
	ADDRESS
WHERE
	POSTAL_CODE = ''
	OR POSTAL_CODE = '35200'
	OR POSTAL_CODE = '17886';

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
6 rows affected.


[(1, '47 MySakila Drive', None, 'Alberta', 300, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (2, '28 MySQL Boulevard', None, 'QLD', 576, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (3, '23 Workhaven Lane', None, 'Alberta', 300, '', '14033335568', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (4, '1411 Lillydale Drive', None, 'QLD', 576, '', '6172235589', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (5, '1913 Hanoi Way', '', 'Nagasaki', 463, '35200', '28303384290', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (6, '1121 Loja Avenue', '', 'California', 449, '17886', '838635286649', datetime.datetime(2006, 2, 15, 9, 45, 30))]

## 문제 8-1

> 문제 8번에 적용된 각 조건마다 컬럼을 할당하고 싶을때 case (python의 if절과 동일) 구문을 사용한다.

case 구문은 
```python
case when 조건1 then action
    else 
    end 
```
와 같이 사용할 수 있다.

- when절 : 조건을 확인하는 구문
- then절 : when절의 조건을 만족했을때 then절에 있는 값으로 컬럼을 채운다.
- else절 : when절의 조건을 만족하지 못했을때 else절에 있는 값으로 컬럼을 채운다.

In [14]:
%%sql

SELECT *,
CASE
	WHEN POSTAL_CODE = '' THEN 'EMPTY'	-- POSTAL_CODE 가 ''(공백)이면 EMPTY 라고 입력
	ELSE POSTAL_CODE	-- ELSE => POSTAL_CODE가 ''이 아니라면 기존 POSTAL_CODE 의 값을 그대로 입력
END AS CLASSIFY_EMPTY_POSTALCODE -- CASE문으로 생성한 데이터가 입력될 컬럼명
FROM
	ADDRESS
WHERE
	POSTAL_CODE = ''
OR POSTAL_CODE = '35200'
OR POSTAL_CODE = '17886';

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
6 rows affected.


[(1, '47 MySakila Drive', None, 'Alberta', 300, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30), 'EMPTY'),
 (2, '28 MySQL Boulevard', None, 'QLD', 576, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30), 'EMPTY'),
 (3, '23 Workhaven Lane', None, 'Alberta', 300, '', '14033335568', datetime.datetime(2006, 2, 15, 9, 45, 30), 'EMPTY'),
 (4, '1411 Lillydale Drive', None, 'QLD', 576, '', '6172235589', datetime.datetime(2006, 2, 15, 9, 45, 30), 'EMPTY'),
 (5, '1913 Hanoi Way', '', 'Nagasaki', 463, '35200', '28303384290', datetime.datetime(2006, 2, 15, 9, 45, 30), '35200'),
 (6, '1121 Loja Avenue', '', 'California', 449, '17886', '838635286649', datetime.datetime(2006, 2, 15, 9, 45, 30), '17886')]

# 문제 9

> ADDRESS 테이블을 이용하여, ADDRESS2 값이 빈값 (NULL) 이거나 35200, 17886에 해당하는 ADDRESS에 모든 정보를 확인해주세요.

In [15]:
%%sql

SELECT *
FROM
	ADDRESS
WHERE
	ADDRESS2 IS NULL
	OR POSTAL_CODE = '35200'
	OR POSTAL_CODE = '17886';

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
6 rows affected.


[(1, '47 MySakila Drive', None, 'Alberta', 300, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (2, '28 MySQL Boulevard', None, 'QLD', 576, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (3, '23 Workhaven Lane', None, 'Alberta', 300, '', '14033335568', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (4, '1411 Lillydale Drive', None, 'QLD', 576, '', '6172235589', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (5, '1913 Hanoi Way', '', 'Nagasaki', 463, '35200', '28303384290', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (6, '1121 Loja Avenue', '', 'California', 449, '17886', '838635286649', datetime.datetime(2006, 2, 15, 9, 45, 30))]

## 문제 9-1
> 위에서 활용한 case문을 적용해서 address2가 NULL인 경우 True, 아닌 경우 False를 입력하도록 지정

In [16]:
%%sql

SELECT 
	*,
	CASE WHEN ADDRESS2 IS NULL THEN 'TRUE'
		ELSE 'FALSE'
	END AS IS_NULL
FROM
	ADDRESS
WHERE
	ADDRESS2 IS NULL
	OR POSTAL_CODE = '35200'
	OR POSTAL_CODE = '17886';

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
6 rows affected.


[(1, '47 MySakila Drive', None, 'Alberta', 300, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30), 'TRUE'),
 (2, '28 MySQL Boulevard', None, 'QLD', 576, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30), 'TRUE'),
 (3, '23 Workhaven Lane', None, 'Alberta', 300, '', '14033335568', datetime.datetime(2006, 2, 15, 9, 45, 30), 'TRUE'),
 (4, '1411 Lillydale Drive', None, 'QLD', 576, '', '6172235589', datetime.datetime(2006, 2, 15, 9, 45, 30), 'TRUE'),
 (5, '1913 Hanoi Way', '', 'Nagasaki', 463, '35200', '28303384290', datetime.datetime(2006, 2, 15, 9, 45, 30), 'FALSE'),
 (6, '1121 Loja Avenue', '', 'California', 449, '17886', '838635286649', datetime.datetime(2006, 2, 15, 9, 45, 30), 'FALSE')]

# 문제 10
> 고객의 성에 'John' 이라는 단어가 들어가는 고객의 이름과 성을 모두 찾아 주세요.

In [17]:
%%sql

SELECT
	FIRST_NAME,
	LAST_NAME
FROM
	CUSTOMER
WHERE
	LAST_NAME LIKE '%John%';

 * postgresql://myuser:***@206.189.155.123:5433/dvdrental
2 rows affected.


[('Patricia', 'Johnson'), ('Kristin', 'Johnston')]